py3status/modules/speedtest.py

```diff
@@ -0,0 +1,298 @@
+    format += '[\?if=download<previous_download ↑ {download} {download_unit}]'
+    format += '[\?if=download=previous_download -> {download} {download_unit}]'
+    format += '|speedtest'
+}
+```
+
+SAMPLE OUTPUT
+[
+    {"full_text": "Speedtest " },
+    {"full_text": "ping ", "color": "#a9a9a9"},
+    {"full_text": "9.498 ms", "color": "#ebdbb2"},
+    {"full_text": " up ", "color": "#a9a9a9"},
+    {"full_text": "78.42 MB/s", "color": "#fb4934"},
+    {"full_text": " down ", "color": "#a9a9a9"},
+    {"full_text": "39.58 MB/s", "color": "#fb4934"}
+]
+"""
+
+from json import loads
+import time
+import threading
+
+STRING_NOT_INSTALLED = "not installed"
+
+
+class SpeedtestCli(threading.Thread):
+    """
+    A thread running speedtest-cli
+    """
+
+    def __init__(self, parent):
+        super(SpeedtestCli, self).__init__()
+        self.parent = parent
+
+    def run(self):
+        self.parent._start()
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    button_share = None
+    format = (
+        u"Speedtest[\?color=darkgray  {elapsed_time} s ]"
+        "[\?color=darkgray  ping [\?color=ping {ping} ms ]"
+        "up [\?color=upload {upload} {upload_unit}] "
+        "down [\?color=download {download} {download_unit}]]"
+    )
+    server_id = None
+    si_units = False
+    thresholds = {
+        "download": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+        "ping": [(200, "bad"), (150, "orange"), (100, "degraded"), (10, "good")],
+        "upload": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+    }
+    timeout = 10
+    unit_bitrate = "MB/s"
+    unit_size = "MB"
+
+    def post_config_hook(self):
+        if not self.py3.check_commands("speedtest-cli"):
+            raise Exception(STRING_NOT_INSTALLED)
+
+        self.button_refresh = 2
+        self.cached_until = self.py3.CACHE_FOREVER
+        self.command = "speedtest-cli --json --secure --timeout {}".format(
+            self.timeout
+        )
+        self.elapsed_time = None
+        self.killed = threading.Event()
+        self.placeholders = self.py3.get_placeholders_list(self.format)
+        self.sleep_timeout = 0
+        self.speedtest_data = {}
+        self.start_time = None
+        self.thresholds_init = self.py3.get_color_names_list(self.format)
+        self.url = None
+
+        # if download* or upload* missing, run complete test
+        if any(
+            key.startswith(x)
+            for x in ["download", "upload"]
+            for key in self.placeholders
+        ):
+            if not any(key.startswith("upload") for key in self.placeholders):
+                self.command += " --no-upload"
+
+            if not any(key.startswith("download") for key in self.placeholders):
+                self.command += " --no-download"
+
+        if self.server_id and int(self.server_id):
+            self.command += " --server " + str(self.server_id)
```

Comment:
We probably don't need this `and int()` check and `"... {}".format(self.server.id)` looks nice (imho).